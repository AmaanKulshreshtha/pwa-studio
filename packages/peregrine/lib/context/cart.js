import React, { createContext, useContext, useEffect, useMemo } from 'react';
import { connect } from 'react-redux';
import { useMutation } from '@apollo/react-hooks';
import gql from 'graphql-tag';

import actions from '../store/actions/cart/actions';
import * as asyncActions from '../store/actions/cart/asyncActions';
import bindActionCreators from '../util/bindActionCreators';
import { BrowserPersistence } from '../util';

const CartContext = createContext();

const isCartEmpty = cart =>
    !cart || !cart.details.items || cart.details.items.length === 0;

const getTotalQuantity = items =>
    items.reduce((total, item) => total + item.quantity, 0);

const createCartMutation = gql`
    mutation createCart {
        cartId: createEmptyCart
    }
`;

const storage = new BrowserPersistence();

function retrieveCartId() {
    return storage.getItem('cartId');
}

function saveCartId(id) {
    return storage.setItem('cartId', id);
}

const CartContextProvider = props => {
    const { actions, asyncActions, cartState, children } = props;

    const [fetchCartId, { data }] = useMutation(createCartMutation, {
        fetchPolicy: 'no-cache'
    });

    // Make deeply nested details easier to retrieve and provide empty defaults
    const derivedDetails = useMemo(() => {
        if (isCartEmpty(cartState)) {
            return {
                currencyCode: 'USD',
                numItems: 0,
                subtotal: 0
            };
        } else {
            return {
                currencyCode: cartState.details.prices.grand_total.currency,
                numItems: getTotalQuantity(cartState.details.items),
                subtotal: cartState.details.prices.grand_total.value
            };
        }
    }, [cartState]);

    const storageCartId = retrieveCartId();

    const derivedCartState = {
        ...cartState,
        isEmpty: isCartEmpty(cartState),
        derivedDetails,
        ...(storageCartId ? { cartId: storageCartId } : {})
    };

    const cartApi = useMemo(
        () => ({
            actions,
            ...asyncActions
        }),
        [actions, asyncActions]
    );

    // cartId stored in local storage and set on cart context (instead of reducer value). Alternatively could use an @client query for the value.
    useEffect(() => {
        if (data && data.cartId) {
            saveCartId(data.cartId);
        }
    }, [cartApi, data]);

    useEffect(() => {
        if (!storageCartId) {
            fetchCartId();
        }
    }, [data, fetchCartId, storageCartId]);

    const contextValue = useMemo(() => [derivedCartState, cartApi], [
        cartApi,
        derivedCartState
    ]);

    return (
        <CartContext.Provider value={contextValue}>
            {children}
        </CartContext.Provider>
    );
};

const mapStateToProps = ({ cart }) => ({ cartState: cart });

const mapDispatchToProps = dispatch => ({
    actions: bindActionCreators(actions, dispatch),
    asyncActions: bindActionCreators(asyncActions, dispatch)
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CartContextProvider);

export const useCartContext = () => useContext(CartContext);
