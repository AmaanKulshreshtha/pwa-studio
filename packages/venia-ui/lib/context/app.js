import React, { createContext, useMemo } from 'react';

import { closeDrawer } from '../actions/app';
import { connect } from '../drivers';

export const AppContext = createContext();

const AppContextProvider = props => {
    const { app: appState, children, closeDrawer } = props;

    const appApi = useMemo(
        () => ({
            closeDrawer
        }),
        [closeDrawer]
    );

    const contextValue = useMemo(() => [appState, appApi], [appApi, appState]);

    return (
        <AppContext.Provider value={contextValue}>
            {children}
        </AppContext.Provider>
    );
};

const mapStateToProps = ({ app }) => ({ app });

const mapDispatchToProps = { closeDrawer };

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(AppContextProvider);
