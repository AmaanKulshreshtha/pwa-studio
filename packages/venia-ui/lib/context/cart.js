import React, { createContext, useMemo } from 'react';

import {
    addItemToCart,
    beginEditItem,
    createCart,
    endEditItem,
    getCartDetails,
    removeItemFromCart,
    updateItemInCart
} from '../actions/cart';
import { connect } from '../drivers';

export const CartContext = createContext();

const CartContextProvider = props => {
    const {
        addItemToCart,
        beginEditItem,
        cart: cartState,
        children,
        createCart,
        endEditItem,
        getCartDetails,
        removeItemFromCart,
        updateItemInCart
    } = props;

    const cartApi = useMemo(
        () => ({
            addItemToCart,
            beginEditItem,
            createCart,
            endEditItem,
            getCartDetails,
            removeItemFromCart,
            updateItemInCart
        }),
        [
            addItemToCart,
            beginEditItem,
            createCart,
            endEditItem,
            getCartDetails,
            removeItemFromCart,
            updateItemInCart
        ]
    );

    const contextValue = useMemo(() => [cartState, cartApi], [
        cartApi,
        cartState
    ]);

    return (
        <CartContext.Provider value={contextValue}>
            {children}
        </CartContext.Provider>
    );
};

const mapStateToProps = ({ cart }) => ({ cart });

const mapDispatchToProps = {
    addItemToCart,
    beginEditItem,
    createCart,
    endEditItem,
    getCartDetails,
    removeItemFromCart,
    updateItemInCart
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CartContextProvider);
