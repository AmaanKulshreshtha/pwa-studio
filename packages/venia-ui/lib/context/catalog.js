import React, { createContext, useMemo } from 'react';

import catalogActions from '../actions/catalog';
import { connect } from '../drivers';

export const CatalogContext = createContext();

const CatalogContextProvider = props => {
    const { catalog: catalogState, children, updateCategories } = props;

    const catalogApi = useMemo(
        () => ({
            updateCategories
        }),
        [updateCategories]
    );

    const contextValue = useMemo(() => [catalogState, catalogApi], [
        catalogApi,
        catalogState
    ]);

    return (
        <CatalogContext.Provider value={contextValue}>
            {children}
        </CatalogContext.Provider>
    );
};

const mapStateToProps = ({ catalog }) => ({ catalog });

const mapDispatchToProps = {
    updateCategories: catalogActions.updateCategories
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CatalogContextProvider);
