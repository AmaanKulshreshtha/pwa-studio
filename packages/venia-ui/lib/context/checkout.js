import React, { createContext, useMemo } from 'react';

import {
    beginCheckout,
    cancelCheckout,
    getShippingMethods,
    resetCheckout,
    submitBillingAddress,
    submitOrder,
    submitPaymentMethod,
    submitPaymentMethodAndBillingAddress,
    submitShippingAddress,
    submitShippingMethod
} from '../actions/checkout';
import { connect } from '../drivers';

export const CheckoutContext = createContext();

const CheckoutContextProvider = props => {
    const {
        beginCheckout,
        cancelCheckout,
        checkout: checkoutState,
        children,
        getShippingMethods,
        resetCheckout,
        submitBillingAddress,
        submitOrder,
        submitPaymentMethod,
        submitPaymentMethodAndBillingAddress,
        submitShippingAddress,
        submitShippingMethod
    } = props;

    const checkoutApi = useMemo(
        () => ({
            beginCheckout,
            cancelCheckout,
            getShippingMethods,
            resetCheckout,
            submitBillingAddress,
            submitOrder,
            submitPaymentMethod,
            submitPaymentMethodAndBillingAddress,
            submitShippingAddress,
            submitShippingMethod
        }),
        [
            beginCheckout,
            cancelCheckout,
            getShippingMethods,
            resetCheckout,
            submitBillingAddress,
            submitOrder,
            submitPaymentMethod,
            submitPaymentMethodAndBillingAddress,
            submitShippingAddress,
            submitShippingMethod
        ]
    );

    const contextValue = useMemo(() => [checkoutState, checkoutApi], [
        checkoutApi,
        checkoutState
    ]);

    return (
        <CheckoutContext.Provider value={contextValue}>
            {children}
        </CheckoutContext.Provider>
    );
};

const mapStateToProps = ({ checkout }) => ({ checkout });

const mapDispatchToProps = {
    beginCheckout,
    cancelCheckout,
    getShippingMethods,
    resetCheckout,
    submitBillingAddress,
    submitOrder,
    submitPaymentMethod,
    submitPaymentMethodAndBillingAddress,
    submitShippingAddress,
    submitShippingMethod
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CheckoutContextProvider);
