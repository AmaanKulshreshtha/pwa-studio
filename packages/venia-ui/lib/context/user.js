import React, { createContext, useMemo } from 'react';

import {
    createAccount,
    getUserDetails,
    signIn,
    signOut
} from '../actions/user';
import { connect } from '../drivers';

export const UserContext = createContext();

const UserContextProvider = props => {
    const {
        children,
        createAccount,
        getUserDetails,
        signIn,
        signOut,
        user: userState
    } = props;

    const userApi = useMemo(
        () => ({
            createAccount,
            getUserDetails,
            signIn,
            signOut
        }),
        [createAccount, getUserDetails, signIn, signOut]
    );

    const contextValue = useMemo(() => [userState, userApi], [
        userApi,
        userState
    ]);

    return (
        <UserContext.Provider value={contextValue}>
            {children}
        </UserContext.Provider>
    );
};

const mapStateToProps = ({ user }) => ({ user });

const mapDispatchToProps = {
    createAccount,
    getUserDetails,
    signIn,
    signOut
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(UserContextProvider);
