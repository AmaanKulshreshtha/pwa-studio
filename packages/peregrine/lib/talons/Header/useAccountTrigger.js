import { useState, useCallback, useEffect } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { useApolloClient, useMutation } from '@apollo/react-hooks';

import { useUserContext } from '@magento/peregrine/lib/context/user';
import { useDropdown } from '@magento/peregrine/lib/hooks/useDropdown';
import { clearCartDataFromCache } from '@magento/peregrine/lib/Apollo/clearCartDataFromCache';
import { clearCustomerDataFromCache } from '@magento/peregrine/lib/Apollo/clearCustomerDataFromCache';

/**
 * The useAccountTrigger talon complements the AccountTrigger component.
 *
 * @returns {Object}    talonProps
 * @returns {Boolean}   talonProps.accountMenuIsOpen - Whether the menu that this trigger toggles is open or not.
 * @returns {Ref}       talonProps.accountMenuRef - A React ref to the menu that this trigger toggles.
 * @returns {Ref}       talonProps.accountMenuTriggerRef - A React ref to the trigger element itself.
 * @returns {Function}  talonProps.handleSignOut - A function for handling sign out.
 * @returns {Function}  talonProps.handleTriggerClick - A function for handling when the trigger is clicked.
 * @returns {Boolean}   talonProps.isUserSignedIn - Whether there is a user currently signed in or not.
 */
export const useAccountTrigger = props => {
    const { mutations, VIEWS } = props;
    const { signOut: signOutMutation } = mutations;

    const [view, setView] = useState(VIEWS.SIGNIN);
    const [username, setUsername] = useState('');

    const {
        elementRef: accountMenuRef,
        expanded: accountMenuIsOpen,
        setExpanded: setAccountMenuIsOpen,
        triggerRef: accountMenuTriggerRef
    } = useDropdown();

    const apolloClient = useApolloClient();
    const history = useHistory();
    const location = useLocation();
    const [revokeToken] = useMutation(signOutMutation);
    const [{ isSignedIn: isUserSignedIn }, { signOut }] = useUserContext();

    const handleSignOut = useCallback(async () => {
        setView(VIEWS.SIGNIN);
        setAccountMenuIsOpen(false);

        // Delete cart/user data from the redux store.
        await signOut({ revokeToken });
        await clearCartDataFromCache(apolloClient);
        await clearCustomerDataFromCache(apolloClient);

        // Refresh the page as a way to say "re-initialize". An alternative
        // would be to call apolloClient.resetStore() but that would require
        // a large refactor.
        history.go(0);
    }, [
        apolloClient,
        history,
        revokeToken,
        setAccountMenuIsOpen,
        signOut,
        VIEWS.SIGNIN
    ]);

    const handleTriggerClick = useCallback(() => {
        // Toggle the Account Menu.
        setAccountMenuIsOpen(isOpen => !isOpen);
        /**
         * When the dropdown closes, the view has to
         * reset to SINGIN view if the user is not
         * signed in.
         */
        if (!isUserSignedIn) {
            setView(VIEWS.SIGNIN);
        }
    }, [setAccountMenuIsOpen, isUserSignedIn, VIEWS]);

    const handleForgotPassword = useCallback(() => {
        setView(VIEWS.FORGOT_PASSWORD);
    }, [VIEWS.FORGOT_PASSWORD]);

    const handleCreateAccount = useCallback(() => {
        setView(VIEWS.CREATE_ACCOUNT);
    }, [VIEWS.CREATE_ACCOUNT]);

    // Close the Account Menu on page change.
    useEffect(() => {
        setAccountMenuIsOpen(false);
    }, [location.pathname, setAccountMenuIsOpen]);

    // Set view to account if the user is signed in and view is not ACCOUNT
    useEffect(() => {
        if (isUserSignedIn && view !== VIEWS.ACCOUNT) {
            setView(VIEWS.ACCOUNT);
        }
    }, [isUserSignedIn, VIEWS, view]);

    return {
        accountMenuIsOpen,
        accountMenuRef,
        accountMenuTriggerRef,
        handleSignOut,
        handleTriggerClick,
        isUserSignedIn,
        handleForgotPassword,
        handleCreateAccount,
        view,
        username,
        setUsername
    };
};
