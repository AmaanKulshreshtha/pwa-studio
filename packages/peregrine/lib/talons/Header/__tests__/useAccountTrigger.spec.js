import React from 'react';

import { useUserContext } from '@magento/peregrine/lib/context/user';
import { useDropdown } from '@magento/peregrine/lib/hooks/useDropdown';

import { useAccountTrigger } from '../useAccountTrigger';
import createTestInstance from '../../../util/createTestInstance';

jest.mock('react-router-dom', () => ({
    useHistory: jest.fn().mockReturnValue({
        go: jest.fn()
    }),
    useLocation: jest.fn().mockReturnValue({
        pathname: '/'
    })
}));

jest.mock('@apollo/react-hooks', () => ({
    useApolloClient: jest.fn().mockReturnValue({}),
    useMutation: jest
        .fn()
        .mockReturnValue([
            jest.fn(),
            { error: { graphQLErrors: [{ message: 'Signout Error' }] } }
        ])
}));

jest.mock('@magento/peregrine/lib/context/user', () => ({
    useUserContext: jest
        .fn()
        .mockReturnValue([{ isSignedIn: false }, { signOut: jest.fn() }])
}));

jest.mock('@magento/peregrine/lib/hooks/useDropdown', () => ({
    useDropdown: jest.fn().mockReturnValue({
        elementRef: 'elementRef',
        expanded: false,
        setExpanded: jest.fn(),
        triggerRef: jest.fn()
    })
}));

jest.mock('@magento/peregrine/lib/Apollo/clearCartDataFromCache', () =>
    jest.fn()
);

jest.mock('@magento/peregrine/lib/Apollo/clearCustomerDataFromCache', () =>
    jest.fn()
);

const defaultProps = {
    VIEWS: {
        SIGNIN: 'SIGNIN',
        FORGOT_PASSWORD: 'FORGOT_PASSWORD',
        CREATE_ACCOUNT: 'CREATE_ACCOUNT',
        ACCOUNT: 'ACCOUNT'
    },
    mutations: {
        signOut: 'signOutMutation'
    }
};

const Component = props => {
    const talonProps = useAccountTrigger(props);

    return <i talonProps={talonProps} />;
};

const getTalonProps = props => {
    const tree = createTestInstance(<Component {...props} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    const update = newProps => {
        tree.update(<Component {...{ ...props, ...newProps }} />);

        return root.findByType('i').props.talonProps;
    };

    return { talonProps, tree, update };
};

test('should return correct shape', () => {
    const { talonProps } = getTalonProps(defaultProps);

    expect(talonProps).toMatchSnapshot();
});

test('handleTriggerClick should set view to SIGNIN if user is not signed in', () => {
    useUserContext.mockReturnValueOnce([
        { isSignedIn: false },
        { signOut: jest.fn() }
    ]);

    const { talonProps, update } = getTalonProps(defaultProps);

    talonProps.handleTriggerClick();

    const updatedTalonProps = update();

    expect(updatedTalonProps.view).toBe('SIGNIN');
});

test('handleForgotPassword should set view to FORGOT_PASSWORD', () => {
    const { talonProps, update } = getTalonProps(defaultProps);

    talonProps.handleForgotPassword();

    const updatedTalonProps = update();

    expect(updatedTalonProps.view).toBe('FORGOT_PASSWORD');
});

test('handleCreateAccount should set view to CREATE_ACCOUNT', () => {
    const { talonProps, update } = getTalonProps(defaultProps);

    talonProps.handleCreateAccount();

    const updatedTalonProps = update();

    expect(updatedTalonProps.view).toBe('CREATE_ACCOUNT');
});

test('should change view to ACCOUNT if user is logged in', () => {
    useUserContext.mockReturnValueOnce([
        { isSignedIn: true },
        { signOut: jest.fn() }
    ]);

    const { talonProps } = getTalonProps(defaultProps);

    expect(talonProps.view).toBe('ACCOUNT');
});

test('setAccountMenuIsOpen should be called with false on mount', () => {
    const setExpanded = jest.fn();

    useDropdown.mockReturnValueOnce({
        elementRef: 'elementRef',
        expanded: false,
        setExpanded,
        triggerRef: jest.fn()
    });

    getTalonProps(defaultProps);

    expect(setExpanded).toHaveBeenCalledWith(false);
});
