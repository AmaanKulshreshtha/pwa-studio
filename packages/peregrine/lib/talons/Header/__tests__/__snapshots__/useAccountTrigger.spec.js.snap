// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correct shape 1`] = `
Object {
  "accountMenuIsOpen": false,
  "accountMenuRef": "elementRef",
  "accountMenuTriggerRef": [MockFunction],
  "handleCreateAccount": [Function],
  "handleForgotPassword": [Function],
  "handleSignOut": [Function],
  "handleTriggerClick": [Function],
  "isUserSignedIn": false,
  "setUsername": [Function],
  "username": "",
  "view": "SIGNIN",
}
`;
